test/success/num.prl

```diff
@@ -15,3 +15,67 @@ Thm NatOne : [
 ] by [
   auto
 ].
+
+Thm Pred : [
+  (-> nat nat)
+] by [
+  {
+    lam a.
+    fresh a', ind <- elim a;
+    [ `zero; auto ];
+    [ hyp a'; auto ];
+  };
+  auto
+].
+
+Extract Pred.
+
+Def Pred2 = [
+   (lam [x] (nat-rec x zero [y z] y))
+].
+
+Extract Pred2.
+
+Thm Prod2Wf : [
+  Pred2 in (-> nat nat)
+] by [
+  fresh n1, n2, n2', ind <- auto;
+  [ `nat ]; // providing the motive
+  auto;
+].
+
+Thm Plus : [
+  (-> nat nat nat)
+] by [
+  {
+    lam a.
+    fresh a', ind <- elim a;
+    [ `(lam [x] x); auto ];
+    [ `(lam [x] (succ (app ,ind x))); auto ];
+  };
+  auto
+].
+
+Extract Plus.
+
+Def Plus2 = [
+  (lam [a]
+   (nat-rec
+    a
+    (lam [x] x)
+    [a' ind] (lam [x] (succ (app ind x)))))
+].
+
+Thm Plus2Wf : [
+  Plus2 in (-> nat nat nat)
+] by [
+  fresh a, b, b', ind <- auto;
+  [ `(-> nat nat) ];
+  auto
+].
```

Comment:
So, not only we can refine stuff to create the addition function, but also we can prove the addition function is well-typed.